### DISCUSSION TITLE
`defer` with `if`

### DISCUSSION THREAD

**Discussion by Feoramund** [odin-lang/Odin#3782] [General]
The potential for a distinction between the following two blocks of code came up recently in https://github.com/odin-lang/Odin/pull/3781#discussion_r1644748888:

```odin
defer if a { /*...*/ }

defer {
	if a { /*...*/ }
}
```

These two work the same at the moment, which is to say that the `if` is only checked when the deferred statement is run, but the first reads as if the block should only be deferred if `a` is true. Should `defer if a ...` raise a warning, or does this case need more documentation?

**Comment by programandala-net** [comment]
Feoramund escribió/skribis/wrote/scrit (2024-06-18T09:59:24-0700):

> ```odin
> defer if a { /*...*/ }
> 
> defer {
> 	if a { /*...*/ }
> }
> ```
> 
> These two work the same at the moment, which is to say that the `if`
> is only checked when the deferred statement is run,

That's the expected behaviour in both cases, which is clear: "A `defer`
statement defers the execution of a statement until the end of the scope
it is in. (…) You can defer an entire block too". Therefore you can
defer an `if` as well.

> but the first reads as if the block should only be deferred if `a` is true.

You are right, but that impression may be caused by reading the code as
plain English.

> Should `defer if a ...` raise a warning, or does this case need more documentation?

I think nothing is needed. In fact, the current [documentation of
`defer`](https://odin-lang.org/docs/overview/#defer-statement) already
includes an example of `defer if` as a deferred block.
